### PR TITLE
Fix the coordinate issue of moving mouse abs on windows

### PR DESCRIPTION
**Pull request by enximi**
Fix the coordinate issue of moving the mouse with absolute coordinates on Windows.

### CODE REVIEW

**Comment by pentamassiv** [comment]
Thank you for your PR.

Unfortunately I don't have a Windows computer available right now. What is the coordinate issue of moving mouse abs? Is the distance not correct? Is it correct for the relative move?

I think it would be best to do the division in i64, round it and then convert it back to i32. Otherwise we needlessly lose some precision there. Also, please mark the function with #[allow(clippy::cast_precision_loss)] to fix the clippy lint.

**Comment by enximi** [comment]
When the MOUSEEVENTF_ABSOLUTE flag is set, some calculations need to be performed on x and y. For details, please refer to Microsoft's [documentation](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mouse_event#remarks).

I have modified the code, so there should be no more precision issues.

**Comment by enximi** [comment]
Yes, when using absolute movement, the position of the mouse does not match the expected position. For example, if there is a 1920*1080 screen and a function is called to move the mouse to coordinates 1920,1080 in absolute mode, the mouse will actually move to the top right corner of the screen. Relative movement is not affected by this issue but it can be influenced by mouse speed and acceleration settings. For more details, please see [https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mouse_event#remarks](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mouse_event#remarks).

File: src/win/win_impl.rs [hunk 1]
```diff
@@ -151,10 +151,20 @@ impl Mouse for Enigo {
...
+            let (w, h) = self.main_display()?;
+            let w = w as i64;
+            let h = h as i64;
+            let x = x as i64;
+            let y = y as i64;
+            // Add w/2 or h/2 to round off
+            // Multiply by 65535 because MOUSEEVENTF_ABSOLUTE flag is set
+            // See https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mouse_event#remarks
+            let x = (x * 65535 + w / 2 * x.signum()) / w;
+            let y = (y * 65535 + h / 2 * y.signum()) / h;
```
Comment (iwanders) [inline src/win/win_impl.rs:164]: The current code is off by one for some pixels, on a 1080p monitor 512,512 results in a cursor position of 511,511 for example. For a width of 1080, the highest reachable pixel is 1079, so 65535 maps to 1079.

I think these two lines should be:
```rust
let y = (65535.0 / (h - 1) as f64) * (y as f64);
let x = (65535.0 / (w - 1) as f64) * (x as f64);
```

This change seems to work for the coordinates I tested; two 1080p monitors side by side, right being primary. Both negative and positive x coordinates exactly achieve the desired value.

Comment (iwanders) [reply src/win/win_impl.rs:164]: Feel free to cherry-pick 6dec1e29b355d7df1f3d4821b90e2204aa5273df which contains this change as well as a comment explaining it.

Comment (enximi) [reply src/win/win_impl.rs:164]: fixed

File: src/win/win_impl.rs [hunk 2]
```diff
@@ -151,10 +151,20 @@ impl Mouse for Enigo {
...
+        let (flags, x, y) = if coordinate == Coordinate::Abs {
+            // 0-screen width/height - 1 map to 0-65535
+            // Add w/2 or h/2 to round off
+            // See https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-mouse_event#remarks
+            let (w, h) = self.main_display()?;
+            let w = w as i64 - 1;
+            let h = h as i64 - 1;
+            let x = x as i64;
+            let y = y as i64;
+            let x = (x * 65535 + w / 2 * x.signum()) / w;
```
Comment (pentamassiv) [inline src/win/win_impl.rs:163]: What is the `+ w / 2 * x.signum()` for? If I want to move the mouse to 0, the result should still be zero, right? With this, will the mouse not be moved to the center of the screen? The coordinate system of `enigo` starts in the top left corner.

I think this should be:

```
let x = (x * 65535) / w;
let y = (y * 65535) / h;
```

**Review by pentamassiv — COMMENTED** [review]
I don't understand the addition of w/2

**Comment by pentamassiv** [comment]
I fixed the formatting for Rust Nightly. Please rebase your commits. You can also squash your commits or I will squash and merge them.

File: src/win/win_impl.rs (hunk 2 above)
Comment (enximi) [reply src/win/win_impl.rs:163]: `let x = (x * 65535 + w / 2 * x.signum()) / w;` = `let x = ((x as f64 * 65535.0) / w as f64).round() as i64;`
This is a technique for rounding using integers. It addresses potential precision loss issues when converting integers to floating-point numbers, hence employing this method.

**Comment by enximi** [comment]
Done.

**Comment by pentamassiv** [comment]
Thank you for your PR. I published an updated version to crates.io

**Pull request merged**